logging: Minor improvement in instance logging

nordic-krch opened the pull request:
When using instance logging (e.g. LOG_INST_INF) first argument is a pointer to the logging object associated with an instance. It is not used if logging is disabled and that can generate 'unused variable' warning when logging is disabled.

Previously logging object was not created when logging was disabled That was done to save memory but because of that object could not be referenced when logging was disabled. Better approach is to create empty array instead of a logging object. It does not increase memory usage but can be referenced and potential compilation warning is suppressed.

nordic-krch committed:
logging: Minor improvement in instance logging

When using instance logging (e.g. LOG_INST_INF) first argument is
a pointer to the logging object associated with an instance. It is
not used if logging is disabled and that can generate 'unused
variable' warning when logging is disabled.

Previously logging object was not created when logging was disabled
That was done to save memory but because of that object could not be
referenced when logging was disabled. Better approach is to create
empty array instead of a logging object. It does not increase memory
usage but can be referenced and potential compilation warning is
suppressed.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

pdgendt commented on include/zephyr/logging/log_instance.h:
```diff
@@ -141,7 +141,7 @@ struct log_source_dynamic_data {
  */
 #define LOG_INSTANCE_PTR_DECLARE(_name)	\
-	IF_ENABLED(CONFIG_LOG, (Z_LOG_INSTANCE_STRUCT * _name))
+	COND_CODE_1(CONFIG_LOG, (Z_LOG_INSTANCE_STRUCT * _name), (int _name[0]))
```
Zero length arrays are a [GNU C extension](https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html), isn't this a portability issue?

cfriedt replied:
Would `__unused` work in the negative case instead of using an array?

nordic-krch replied:
`__unused` might work but won't be user friendly. Use case is following: There is a object structure that has additional `log` field that is passed to the instance logging (e.g. `LOG_INST_INF(my_obj->log, "log")`). If logging is disabled then structure does not have this element at all (`main` state) so it is not accessed thus if i have a function where i use `my_obj` only to get that log:
``` 
struct_my_obj * my_obj = dev;
LOG_INST_INF(my_obj->log, "log");
```
Then i get warning when it is unused. To avoid warning I would need to add `__usued` to that variable declaration which is not user friendly since user sees that variable is being used so `__usued` requirement seems dubious.

@pdgendt as for portability issue: ZL arrays are used in zephyr (`hci_types.h`, `net_ip.h`). I will try to check if i can add a compile time detection if ZLA is supported and if not use `_name[1]` instead so the cost for other compilers will be that structure will be probably one word bigger.

nordic-krch committed:
toolchain: Add flag indicating zero length array support

Add flag which indicates if a toolchain supports ZLA (Zero
Length Arrays). It is a GCCC extension inherited by clang as
well.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>
